### DISCUSSION TITLE
Mileage tracking (business vs. personal)

### DISCUSSION THREAD

**Discussion by tonypilborg** [teslamate-org/teslamate#999] [Ideas]
**What would you like to be added**:
Ability to flag each drive - as personal or business.
Ability to export reports - with relevant selectable drives data, like start address, end- address, Km, date+time, odometer start-end, etc...
(or more applicable: create a slightly altered Grafana view, with filter for "trip type", and the data from above.. - similar to "Drives" or "Trip" - only with some different data and filters...)

**Why is this needed**:
For tax purposes - and to not having to rely of manual tracking or expensive (or poor) app, with questionable impact on battery of phone etc.

**Additional context**:
It would be brilliant if it was somehow possible to select "something" in the car before or during a drive (I know driver profile is not an option) - perhaps a "mode switch", through MQTT or web (Siri shortcut or similar).. But obviously needed to be able to change/re-classify afterwards.
Perhaps also an option to add a note - like costumer name if parking is not next to that company... I know some of it could be solved with geo-fencing, but perhaps it is not a good idea to abuse to core geofencing for that (so a copy of the function, just for this purpose) - perhaps it is all perfect to use the existing geo-fences??? That could solve automatic "costumer tagging", but a way to alter for a specific drive would be valuable...

**Comment by roadrash2108** [comment]
This seems to be a mix of #505 and #989 and a new feature ("in car tagging")

**Comment by tonypilborg** [comment]
> This seems to be a mix of #505 and #989 and a new feature ("in car tagging")

#989 should be dead easy, since the drives tables has "start_km" and "end_km" - sp that is purely a bit of Grafana exercise..
#505 has some of the same ideas, although I would prefer not to have automation based on any of these existing data (but can live with going through my drives to adjust the type for those that shouldn't stay default)..

But yes - I think either 2 more columns to drives table (the type, could be integer/smallint, and the "note" a varchar or text) - or a separate table referencing the drives table... Both should be a breeze to build - and creating a Grafana layout with all those filtered drives (and perhaps a simple summary on top) also a small feat....

I will gladly participate...... Only challenge I can see, would be to allow for setting a default/automated drive type, but that could be "on top", the data structure could come first, and would give 99% of the benefit...

**Comment by roadrash2108** [comment]
#989 is already done (or at least for a special case), so "only" the part with the data-layer is missing. I would volunteer for the Grafana-part in case someone (@tonypilborg ?) volunteers for the data-layer-part...

**Comment by tonypilborg** [comment]
> #989 is already done (or at least for a special case), so "only" the part with the data-layer is missing. I would volunteer for the Grafana-part in case someone (@tonypilborg ?) volunteers for the data-layer-part...

From my brief (very brief) scanning of the issue, I don't think it is covered - even the special case..
The issue is, that - as has been mentioned in #912 - sometimes it takes quite a while (for my car yesterday almost 15 minutes!) before the api has contact to the car... And that is significant for tax putpose, since end_km for drive x-1 is not (In TeslaMate, but obviously in reality) equal to start_km for drive x...

So for the sake of taxation - it is either (probably both) necessary to be able to adjust manually (although this could be done in an excel extract, or be able to use end_km as the start_km for the next drive.... And it is then complicating the SELECT in the "tax report", since start_km drive x has to be inherited from end_km drive x-1, even if drive x-1 is not included in the select (if it was a private putpose, and thus not applicable for the tax report)...

(in other words - when your are looking at the Grafana part, you should SELECT all drives, but only display those with the chosen "purpose". - and then actually always use end_km from the previous drive as the start_km.... (except if the previous drives wasn't logged - which could happen.. So perhaps include all 3: end_km(x-1), start_km(x), end_km(x) - for all drives, and then filter which lines are shown...)

I actually don't have a dev install of TeslaMAte - but the structure could simply be:

CREATE TABLE public.drivepurpose (
    id integer NOT NULL,
    drive_id integer,
    purpose_id integer,
    customer character varying(255),
);

CREATE TABLE public.purpose (
    id integer NOT NULL,
    purpose_text character varying(255),
);

I would ever only use business or private purposes - but other might choose something more granulated (especially if somebody taskes up the idea of pre-populating the purpose automatically... And if purpose is something other that business, perhaps the drivepurpose_customer should be called something else.... ...

but obviously it is needed to allow maintenance of these tables - links in Grafana to TeslaMate???

**Comment by ngardiner** [comment]
I don't think we'd need a specific column for purpose if tagging (https://github.com/adriankumpf/teslamate/issues/505) was implemented. It would allow you to tag drives with a purpose, but also with many other things (perhaps driver, trip for drives as part of a larger trip, etc). If tagging is implemented as an opt-in system such as charge costs are, where certain geofences might trigger an automatic tag if you configure it or otherwise tagging can be applied manually, it should suit both use cases - those who want automatic tagging or those who want to do it manually.

**Comment by tonypilborg** [comment]
> I don't think we'd need a specific column for purpose if tagging (#505) was implemented. It would allow you to tag drives with a purpose, but also with many other things (perhaps driver, trip for drives as part of a larger trip, etc). If tagging is implemented as an opt-in system such as charge costs are, where certain geofences might trigger an automatic tag if you configure it or otherwise tagging can be applied manually, it should suit both use cases - those who want automatic tagging or those who want to do it manually.

I idsagree.... Or rather, I agree only if tagging could then be more than one dimension... If you consider the case with business trips, a tagging as business makes sense, but it could be very valuable with a way to mark a drive with a customer name (for tax purposes)...
But when pulling out to overview of drives, I would not want to select all customers - I would need to select private vs. business...

But I agree - much more elegant as an opt-in option, like for charging... Although that implementation can not be directly copy-pasted. We would need a table for these tags, irrespective of the number of tagging dimension 👍 

- which will be more complex, that is why I suggested the "purpose" approach....

**Comment by ngardiner** [comment]
Just tag it with the customer's name? Your column is one-dimensional - it's one value per drive. Tagging is inherently multi-dimensional as you can apply multiple tags for multiple things

Frankly, you're pushing for a worse design

**Comment by tonypilborg** [comment]
You are absolutely right!
I didn't even consider a 1:many relation between drives and tags...

That is clearly a good solution - and I can't think of a scenario where some Grafana logic wouldn't be able to cater for any requests/filtering.

Do you have any insights in to the issue with "missing start of drives"? I am aware I could export all to excel, and then do some managing there . but it would be great to have it built in to the report set :) 
(or does this actually belong to #989 ???)

**Comment by ngardiner** [comment]
There are two separate issues covering that, I have been following the issues but I think they're somewhat multi-faceted. The issues are https://github.com/adriankumpf/teslamate/issues/912 and https://github.com/adriankumpf/teslamate/issues/785 and I do think the suggestion of a UI to allow modifying the start/end time of a drive is a good one, perhaps one to propose in #912

**Comment by emilfors** [comment]
You're welcome to check out my tesla_journal repo (https://github.com/emilfors/tesla_journal) if you need this functionality ASAP. The UI is in Swedish only so far, so sorry about that. Work in progress.

**Comment by tonypilborg** [comment]
> You're welcome to check out my tesla_journal repo (https://github.com/emilfors/tesla_journal) if you need this functionality ASAP. The UI is in Swedish only so far, so sorry about that. Work in progress.

A few notes of feedback - after I have installed:

1. you should mention in the notes, that if Teslamate is run from the docker-compose (which I assume many are doing, especially those that are less experienced), then postgres should expose the port 5432 (you mention "if on another host".
2. Swedish not a problem for me :)
3. for the purpose of tax, a few things would need a change (I assume for most tax-offices):
3.a. You need from a to adresses - not TeslaMate "location names"
3.b. You probably need start and end odometer (please see note above regarding using end_km from previous drive as start_km)
3.c. It is most likely needed to be able to export.... And perhaps consider reversing the order (both dates and start above end for each drive)
3.d. For a report (export or print) it would be beneficial to have the option to hide private or business trips - and only include those relevant... (it is just for me to know where I drive privately - neither my boss nor tax needs to have that detailed info on this :) 

And some "bugs" (I assume):
1. I have data since February 2020 - but I get empty reports for February, March, April and September... No pattern there, so not sure if some bad data (missing end of drives or similar) could break a month report...
(upd: yes - some nil data from end_date..)

And then some of the generic notes:

I like some of the ideas - the grouping/ungrouping is great - and marking of private vs. business equally so, It is super easy to use!
But I am confused why you decided to go this route - and not work within Grafana and the exisitng project?? I assume you maintain your own internal database for the classification and grouping of trips - and haven't added tables to the existing database?? Where is that stored?

Definitely some good ideas - let me know if you need inputs (I can translate to Danish and English with confidence) or testing....

**Comment by tonypilborg** [comment]
> And then some of the generic notes:
> 
> I like some of the ideas - the grouping/ungrouping is great - and marking of private vs. business equally so, It is super easy to use!
> But I am confused why you decided to go this route - and not work within Grafana and the exisitng project?? I assume you maintain your own internal database for the classification and grouping of trips - and haven't added tables to the existing database?? Where is that stored?
> 
OK - I just browsed the code, and I can see you have added tables in the TeslaMate database... That worries a bit, and I would have preferred a note about that. (but no panic, I assume you are only writing to those 2 tables?

But for the sake of sleeping calm - I would probably go with separate db users for TeslaMate and Tesla_Journal, and make sure the latter can only read from the main tables... But anyway some noise also regarding backup etc...
I really hope you wish to join your ideas in to the main project?? Or is there a compelling reason your would prefer an external program??

**Comment by emilfors** [comment]
Hi Tony, thanks for checking out the project. You're actually its very first user besides me :)

Here's a brief background which I guess will answer a few of your questions. I took delivery of my Tesla late August. I used to have an OBDII dongle-based solution for keeping a drive journal, but soon realized that wouldn't fly with the new car. So I started looking for alternatives. While doing that I came across TeslaMate which seemed like an awesome thing altogether and also like it might suit my journal keeping needs. Unfortunately it didn't, but I thought I could at least leverage its data tracking and tweak the Grafana UI into doing what I needed. So I started trying to do that, I think I spent a couple of weeks on it but eventually gave up. I just couldn't seem to understand how to make Grafana accept input and not just produce data visualizations. I toyed with the idea of modifying TeslaMate itself but that would still mean I'd need to build a separate UI (since I couldn't make Grafana act as one).
Meanwhile, I had to learn Go for a new project at work and so I figured what better way to do that than to build something useful, right? And because I didn't feel like learning two new languages at once (I don't know Elixir at all) I dropped the thought of doing anything with TeslaMate directly. Hence this project.

This wasn't intended for anyone else's use but mine, which may or may not be obvious. That's one reason why I didn't actually consider security at all (i.e. just piggybacking on the teslamate db user). Another is that I didn't know PostgreSQL either and didn't want to spend time figuring out users, access rights and all that. I thought IF that would prove necessary, then that's always something that can be changed later. To answer your question, yes, my project writes two its own two tables only, and reads from the others.

I made the repo public because I thought it might actually be useful to someone else and because I could use some help testing it. As I made this for my own intents and purposes, it caters to Swedish regulations only at this point - as far as I know it's compliant barring the fact that you need to be able to add comments to the trips (so you can let the tax authorities know what they were for). That's what I'm working on now - adding a detailed trip view where you can add that information. After that, the next step is implementing a report generator (probably just spitting out a PDF). And just as you say, in the generated report I intend to include odometer readings, hide the private trips and so on.

TL;DR - I made this in part because I need it personally and in part as a way of teaching myself Go, PostgreSQL, JQuery and a few other things that I needed to learn anyway.

**Comment by tonypilborg** [comment]
@emilfors:
Makes sense. I really like your take on grouping and marking (business vs. private) - would actually consider using it, and then modify the standard "Dutch tax" report to include/consider/filter based on the marking made in Tesla_Journal....
(but at the same time, I realized that this "Dutch-tax" solves 98% of my need for now - not as elegant as if the classification/filter is done in TeslaMate+Grafana - but doable in Excel :)

And the option to set the type of drive - the default marking - through a webservice (sure it would be easy in Tesla_journal) - so that I would have to think about it, if I just remember to set the proper type before driving - that would be great :)
(and most likely force me to open my TeslaMate installation to the internet - currently it is only addressable on my LAN :)

**Comment by tonypilborg** [comment]
@emilfors:

I finally got my finger out of wherever i forgot it....... And maintained my 4 corrupted drives in the database...

After having made my first month of mileage reporting - in excel with a dump from the "dutch_tax" report - I realize how much of an improvement the classification AND grouping of drives could be.

A few suggestion:
0. Possibility to replace an address with a custom address (I can possibly solve through geofences in TeslaMate.)
1. A custom date interval, not just full months (I need to report from 21st through the 20th the following month)
2. Possibility to only show drives with a specific classification (none, business or personal) - a crude filter.
2.b. I realize you have "printing to PDF" on the to-do, and it would make sense in that - but perhaps also in view itself??
3. Translation of labels.....
4. Start and end odometer.... Just because it is probably a good idea to give tax auths even more data :)
5. A comment of each drive (like purpose or customer)

I REALLY like hos the grouping of trips is designed - easy and efficient. - correct me if I am wrong, but you just use:
Start address and time of first drive in a group - and end address and time of the last drive.
adding the drive time and km (not just from start_odo of first through end_odo of last)
I am not entirely sure how to setup this behavior in Grafana... But perhaps I will attempt on a copy...

@adriankumpf: it would be SO nice, if you could make a switch in the TeslaMate interface - to select default classification of drives (business vs. personal)... Perhaps complying with the table name introduced by Emil??

**Comment by emilfors** [comment]
Hey Tony, sorry for the late reply. Totally swamped with "real" work at the moment so there hasn't been much time to work on this lately. I'm hoping to change that soon :)
As for your suggestions:
0. Yeah I haven't bothered doing anything with addresses and such since that can be done fairly simply through TeslaMate / Grafana.
1. I hadn't thought of that, I've always reported full calendar months. Shouldn't be a hard thing to fix of course.
2. Yeah, a filter would be easy enough to implement too of course. I had already planned to include business trips only in the generated report.
3. Heh. You mean not everyone speaks Swedish? :D
4. The odometer I thought would just clutter the presentation up, but I am planning to include them in the report (precisely for the reason you mentioned)
5. Yes, comments are what I'm working on now. Required by the Swedish tax authority.

I'm glad you like the grouping feature! I think it's pretty useful myself. Yes, that's exactly right. Start time and address of the first drive in the group, end time and address of the last and the driven distance and time are the sums of all drives in the group.

Cheers :)

**Comment by TMahlmann** [comment]
@adriankumpf could we potentially get an update on this feature request to sort trips between two locations or trips that begin from certain destinations in our TeslaMate as business / personal? With tax day here in the United States coming up on April 15 -- it would be so amazing to have this feature instead of going through and manually classifying every trip we've got logged in our drives.

**Comment by TMahlmann** [comment]
@adriankumpf still looking to see if this feature request could get some attention with Tax day coming up here pretty fast here in the United States. Would it be possible to add these business/personal mileage tracking features to Teslamate/Grafana?

**Reply by babatoko** [reply]
Would be nice to get a return on this or any request and not ignore them.

Just a Nope won't do that is still a valid return.

Thanks

**Reply by tonypilborg** [reply]
I choose to use Tesla_Journal to classify all drives - end then I just modified the “Dutch Tax” report to only include those drives classified as business.....

quite simple - all should be able to do this :)

**Reply by Knuppel1983** [reply]
> I choose to use Tesla_Journal to classify all drives - end then I just modified the “Dutch Tax” report to only include those drives classified as business.....
> 
> quite simple - all should be able to do this :)

Can you explain how you did this? I have no idea what 'Go toolchain' is. I have a working Teslamate setup running in Docker (Synology). How do i proceed from here?

**Comment by PCjabber** [comment]
I would also love to see some form of this implemented (#1258 seems to be a similar idea), or even just a comments/notes field for Drives as #2492 describes.
